### DISCUSSION TITLE
TwinBasic

### DISCUSSION THREAD

**Discussion by Docent24** [cristianbuse/VBA-FastDictionary#9] [General]
Hi. Is this project compatible with TwinBasic? And if not, do you plan to make it so?

Thanks.

**Comment by GCuser99** [comment]
I tried v2.1.0 in tB a few days ago and there were no errors on import. The demo worked until the two enumerations at the end. The first failed with an error and the second did not do anything. So it is getting close... I'm going to wait and try again after the next release of tB. Then I'll get serious about trying to isolate/report the problem(s).

**Comment by cristianbuse** [comment]
Hi,

I would be happy to make this TB compatible. However, it is my understanding that Wayne already built a TB native Dictionary which is surely faster since it compiles unlike mine.

Is there any benefit in doing this?

**Comment by GCuser99** [comment]
As far as I know tB does not have its own native Dictionary (maybe you are thinking of the native Collection?). I would use your Fast-Dictionary once it is working in tB (again I'm gonna wait until next big, anticipated update to tB for testing). My long-term goal is to reduce/eliminate associated VBScript dependencies.

**Comment by cristianbuse** [comment]
Hi @GCuser99 ,

Are you aware of any conditional compilation or statement in TB that will allow me to distinguish between compiled and non-compiled?
I ask because, if compiled, there will be no need for using the hashing of Scripting.Dictionary for strings and I could just skip that code which would make the dictionary faster. For non-compiled, I would keep it. I haven't looked into TB for the last 2 years but in my understanding, the code could be uncompiled depending on licence. Please correct me if I am wrong.

Thanks!

**Comment by GCuser99** [comment]
I checked the list of tB compiler constants and it looks like not. But I think `App.IsInIDE` should work at runtime.

**Comment by GCuser99** [comment]
Oh, and yes you are correct on the "uncompiled depending on license" - Wayne has proof-of-concept-ed tBA already so it's coming after V1.

**Comment by cristianbuse** [comment]
@GCuser99 

Thanks for answering, I wasn't aware of the ```App.IsInIDE```. I was planning to use this:
```
'https://stackoverflow.com/a/9516565/8488913
Public Property Get InIde() As Boolean
    Debug.Assert SetTrue(InIde)
End Property
Private Function SetTrue(ByRef b As Boolean) As Boolean
    b = True
    SetTrue = True
End Function
```

Anyway, I've done some testing today and unfortunately, TB erases the static memory structs that are within the global Dictionary instance. My guess is that TB simply ignores the ```FADF_AUTO Or FADF_FIXEDSIZE``` value in ```SAFEARRAY.fFeatures```. Will investigate more in the following days but this might require that I stop using fake arrays to copy memory and use the built-in memory functions in TB. There is a chance that this won't work either and that the global structs need to be replicated per instance (e.g. bit lookups), which will impact performance and memory used. Plus, I have no idea how long this would take me to implement as it would require extensive changes. Will post updates in this thread as I go along.

**Comment by cristianbuse** [comment]
0ece01d5297f2e9037ea6b0ad2dc92df964da83e adds TWINBASIC support. Wiill create a new release soon. Many thanks!

**Reply by cristianbuse** [reply]
18ab788e5d566f7c7defa8c45416f46faffc78a6 this fixes a small bug

**Comment by cristianbuse** [comment]
@GCuser99

Hi Mike, if I were to do a TB-specific version of this Dictionary inside a .twin module instead of .cls, how would that affect you? See #34

**Comment by GCuser99** [comment]
Not much of an impact for me - thanks for asking. I've incorporated Fast-Dictionary in only one (new) project, which is working fine but I don't think I tested "for each". I held back in incorporating in older projects such as our SeleniumVBA COM DLL because some methods in that object model pass Dictionary objects to VBA and so mindfully did not want to mess up anyone's code that uses that DLL with MS Scripting Dictionary.

BTW, I prefer using .twin files in tB. You can place multiple classes and modules in a single .twin file - very convenient for keeping multi-class/module support functionality from cluttering up the project tree.

Thanks Cristian!

**Reply by cristianbuse** [reply]
Yeah, I was thinking a .twin file is the best approach as that would completely remove the need for hacks. But just wasn't sure how would that affect anyone. Thanks a lot Mike!
